File: VB/ColoredProgressBar/CustomPaintedProgressBarHelper.vb
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports System
Imports System.Drawing

Namespace ColoredProgressBar

    Public Class CustomPaintedProgressBarHelper

        Private _Column As GridColumn

        Private _View As GridView

        Public Sub New(ByVal column As GridColumn)
            _Column = column
            _View = TryCast(_Column.View, GridView)
            AddHandler _View.CustomDrawCell, New DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventHandler(AddressOf View_CustomDrawCell)
        End Sub

        Private Sub View_CustomDrawCell(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs)
            If e.Column Is _Column Then
                DrawProgressBar(e)
                e.DefaultDraw()
                e.Handled = True
            End If
        End Sub

        Private Sub DrawProgressBar(ByVal e As DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs)
            Dim percent As Integer = Convert.ToInt16(e.CellValue)
            Dim v As Integer = Convert.ToInt32(e.CellValue)
            v = v * e.Bounds.Width \ 100
            Dim rect As Rectangle = New Rectangle(e.Bounds.X, e.Bounds.Y, v, e.Bounds.Height)
            Dim b As Brush = Brushes.Green
            If percent < 25 Then
                b = Brushes.Red
            ElseIf percent < 50 Then
                b = Brushes.Orange
            ElseIf percent < 75 Then
                b = Brushes.YellowGreen
            End If

            e.Cache.FillRectangle(b, rect)
        End Sub
    End Class
End Namespace

File: CS/ColoredProgressBar/CustomPaintedProgressBarHelper.cs
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System;
using System.Drawing;

namespace ColoredProgressBar {
    public class CustomPaintedProgressBarHelper {
        GridColumn _Column;
        GridView _View;
        public CustomPaintedProgressBarHelper(GridColumn column) {
            _Column = column;
            _View = _Column.View as GridView;
            _View.CustomDrawCell += new DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventHandler(View_CustomDrawCell);
        }
        private void View_CustomDrawCell(object sender, DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs e) {
            if (e.Column == _Column) {
                e.DefaultDraw();
                DrawProgressBar(e);
                DrawEditor(e);
                e.Handled = true;
            }
        }
        void DrawProgressBar(DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs e) {
            int percent = Convert.ToInt16(e.CellValue);
            int v = Convert.ToInt32(e.CellValue);
            v = v * e.Bounds.Width / 100;
            Rectangle rect = new Rectangle(e.Bounds.X, e.Bounds.Y, v, e.Bounds.Height);
            Brush b = Brushes.Green;
            if (percent < 25)
                b = Brushes.Red;
            else if (percent < 50)
                b = Brushes.Orange;
            else if (percent < 75)
                b = Brushes.YellowGreen;
            e.Cache.Paint.FillRectangle(e.Cache.Graphics, b, rect);
        }
        void DrawEditor(DevExpress.XtraGrid.Views.Base.RowCellCustomDrawEventArgs e) {
            GridCellInfo cell = e.Cell as GridCellInfo;
            Point offset = cell.CellValueRect.Location;
            BaseEditPainter pb = cell.ViewInfo.Painter as BaseEditPainter;
            if (!offset.IsEmpty)
                cell.ViewInfo.Offset(offset.X, offset.Y);
            try {
                pb.Draw(new ControlGraphicsInfoArgs(cell.ViewInfo, e.Cache, cell.Bounds));
            }
            finally {
                if (!offset.IsEmpty)
                    cell.ViewInfo.Offset(-offset.X, -offset.Y);
            }
        }
    }
}
